src/query/storage/m3/consolidators/fetch_result_map.go

```diff
@@ -0,0 +1,162 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package consolidators
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	xerrors "github.com/m3db/m3/src/x/errors"
+)
+
+type fetchResultMapWrapper struct {
+	resultMap *fetchResultMap
+
+	multiErr xerrors.MultiError
+}
+
+func (w *fetchResultMapWrapper) err() error {
+	return w.multiErr.FinalError()
+}
+
+func (w *fetchResultMapWrapper) len() int {
+	return w.resultMap.Len()
+}
+
+func (w *fetchResultMapWrapper) list() []multiResultSeries {
+	result := make([]multiResultSeries, 0, w.len())
+	for _, results := range w.resultMap.Iter() {
+		result = append(result, results.value)
+	}
+
+	return result
+}
+
+func (w *fetchResultMapWrapper) get(
+	it encoding.SeriesIterator,
+) (multiResultSeries, bool) {
+	return w.resultMap.Get(it)
+}
+
+func (w *fetchResultMapWrapper) set(
+	it encoding.SeriesIterator, series multiResultSeries,
+) {
+	// NB: no need to copy key; the
+	w.resultMap.SetUnsafe(it, series, fetchResultMapSetUnsafeOptions{
+		NoCopyKey:     true,
+		NoFinalizeKey: true,
+	})
+}
+
+// newFetchResultMap builds a MultiFetchResultMap, which is primarily used
+// for checking for existence of particular ident.IDs.
+func newFetchResultMap(size int) *fetchResultMapWrapper {
+	wrapper := &fetchResultMapWrapper{}
+
+	addErr := func(msg string, err error) {
+		err = fmt.Errorf("err, msg: %s, inner: %v", msg, err)
+		wrapper.multiErr = wrapper.multiErr.Add(err)
+	}
+
+	resultMap := _fetchResultMapAlloc(_fetchResultMapOptions{
+		hash: func(it encoding.SeriesIterator) fetchResultMapHash {
+			hash, err := it.Tags().Hash()
+			if err != nil {
+				addErr("hashing iterator failed", err)
+				return 0
+			}
+
+			return fetchResultMapHash(hash)
+		},
+		equals: func(x, y encoding.SeriesIterator) bool {
+			// NB: fail fast on errors.
+			err := wrapper.multiErr.FinalError()
+			if err != nil {
+				return false
+			}
+
+			// NB: succeed fast if possible with matching IDs.
+			if x.ID() == y.ID() {
```

Comment:
Hm, think you want `x.ID().Equal(y)` yeah?

Reply:
Also, wonder if we want to bother doing this check? It might be expensive to always be doing a bytes.Compare(...) and a tag comparison each time?

Reply:
The bytes.Compare is much cheaper than reading all the tags, so better to short circuit early if IDs match (which is the case almost all of the time)